crates/ide-diagnostics/src/handlers/useless_braces.rs

```diff
@@ -15,20 +15,23 @@ pub(crate) fn useless_braces(
 ) -> Option<()> {
     let use_tree_list = ast::UseTreeList::cast(node.clone())?;
     if let Some((single_use_tree,)) = use_tree_list.use_trees().collect_tuple() {
+        // If there is a `self` inside the bracketed `use`, don't show diagnostic.
+        if single_use_tree.syntax().first_token().unwrap().kind() == SyntaxKind::SELF_KW {
+            return Some(());
+        }
```

Comment:
One more thing: `use a::{b::self};` bypasses this check and we'd provide a quickfix that turn it into `use a::b::self;`, which is invalid. Something like [this line](https://github.com/rust-lang/rust-analyzer/pull/15216/files#diff-0cd524758b7692ad400e6cf58628a1016cbd79bcad5e0e06cfdeb50dd7286356L53) from the removed `fn remove_braces()` should handle this kind of cases well.

Reply:
is `use a::{b::self};` a valid syntax? because i got a syntax error

`Syntax Error: The self keyword is only allowed as the first segment of a path` 

when i created a test case

```rust
use a::{b::self};

mod a {  pub mod b { } }
```

Is there is a think that i am missing?

Reply:
Yeah I've noticed that too. Turns out that's rust-analyzer's bug, because rustc accepts it ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=74dd7261a8a50716049ca4aa9d73346f)).

Reply:
We can cover it by disabling `syntax-error` and using `check_diagnostics_with_config`